### PR TITLE
CRM-21677 - Report improvements

### PR DESCRIPTION
**Pull request by yashodha**
Overview
----------------------------------------
Clean up reports that have unnecessary joins to phone, email and address.

Before
----------------------------------------
unnecessary joins to phone, email and address tables

After
----------------------------------------
join only when needed.

---

 * [CRM-21677: Report improvements](https://issues.civicrm.org/jira/browse/CRM-21677)

### CODE REVIEW

**Comment by yashodha** [comment]
test this please

File: CRM/Report/Form/Contribute/Summary.php [hunk 1]
```diff
@@ -41,7 +41,7 @@ class CRM_Report_Form_Contribute_Summary extends CRM_Report_Form {
   protected $_customGroupExtends = array('Contribution', 'Contact', 'Individual');
   protected $_customGroupGroupBy = TRUE;
 
-  public $_drilldownReport = array('contribute/detail' => 'Link to Detail Report');
```
Comment (eileenmcnaughton) [inline CRM/Report/Form/Contribute/Summary.php]: I think this should be in a separate PR as it is a behaviour change & we need to check if it will get buy in (I tend to agree with it). The other changes in this are just technical so better to put on 2 tracks

**Comment by eileenmcnaughton** [comment]
@yashodha this looks like a nice clean up! Yay + double yay.

I think we should probably sync the approach here with that in the extended report extension since we are cleaning up. I had been porting a few of the approaches I use in that extension to core over time but it takes a little co-ordination because if the signature changes in core I have to change it in extended report ( I try to just remove the function & rely on core where I can)

In extended report I use the following:

```
  /**
  * Add join from contact table to email.
   *
   * Prefix will be added to both tables as it's assumed you are using it to get address of a secondary contact.
   *
   * @param string $prefix
  *  @param array $extra Additional options.
  *      Not currently used in core but may be used in override extensions.
   */
  protected function joinEmailFromContact($prefix = '', , $extra = array()) {

   if ($this->isTableSelected($prefix . 'civicrm_email')) {
      $this->_from .= " 
       LEFT JOIN civicrm_email {$this->_aliases[$prefix . 'civicrm_email']}
      ON {$this->_aliases[$prefix . 'civicrm_email']}.contact_id = {$this->_aliases[$prefix . 'civicrm_contact']}.id
     AND {$this->_aliases[$prefix . 'civicrm_email']}.is_primary = 1
";
  }
```

The differences are the parameters that are passed in & the function name. Because I have similar functions for a lot of joins I use the pattern Join...From - since there might also be a join to email from LocationBlock.

Passing in $prefix permits 2 joins to email table (from contact a & from contact b). The $extra array is a relatively recent one & I haven't used it much but I think for a standard pattern it's better to have it - it might be used to indicate 'is_primary_only' = FALSE  in the above for a report that wants all addresses returned.

File: CRM/Report/Form/Contact/Summary.php [hunk 2]
```diff
@@ -193,19 +193,8 @@ public function from() {
...
-    }
-
-    if ($this->_phoneField) {
-      $this->_from .= "
-            LEFT JOIN civicrm_phone {$this->_aliases['civicrm_phone']}
-                   ON {$this->_aliases['civicrm_contact']}.id = {$this->_aliases['civicrm_phone']}.contact_id AND
-                      {$this->_aliases['civicrm_phone']}.is_primary = 1 ";
-    }
+    $this->addPhoneFromClause();
+    $this->addEmailFromClause()
```
Comment (seamuslee001) [inline CRM/Report/Form/Contact/Summary.php]: missing a `;` here @yashodha

Comment (yashodha) [reply CRM/Report/Form/Contact/Summary.php]: @seamuslee001 : fixed

**Comment by eileenmcnaughton** [comment]
test errors are on the contact/detail report

**Comment by yashodha** [comment]
@eileenmcnaughton : the functions now have parameters in the same format.

**Comment by eileenmcnaughton** [comment]
Thanks @yashodha  - are you OK changing the names too? I think it is clearer saying what table it's coming from

**Comment by yashodha** [comment]
@eileenmcnaughton yes, sure!

**Comment by yashodha** [comment]
test this please

**Comment by yashodha** [comment]
test this please

**Comment by yashodha** [comment]
@eileenmcnaughton : made the changes

**Comment by yashodha** [comment]
test this please

File: CRM/Report/Form/Contribute/Summary.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Report/Form/Contribute/Summary.php]: @yashodha this line is the only thing that I think is blocking merge here - because I think it;s a separate change -
 although I think it would be better if we changed joinAddressFromClause to joinAddressFromContactClause - and the same on the other functions. (that would make it more in line with extendedReport and differentiate between joinAddressFromLocation from joinAddressFromContact.

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Report/Form.php [hunk 3]
```diff
@@ -4304,33 +4304,82 @@ public function fiscalYearOffset($fieldName) {
...
+   * Add Country into From Table if required.
+   *
+   * Prefix will be added to both tables as
+   * it is assumed you are using it to get address of a secondary contact.
+   *
+   * @param string $prefix
+   * @param array $extra Additional options.
+   *      Not currently used in core but may be used in override extensions.
+   */
+  public function joinCountryFromClause($prefix = '', $extra = array()) {
```
Comment (eileenmcnaughton) [inline CRM/Report/Form.php]: Can we change this to joinCountryFromAddressClause

File: CRM/Report/Form.php [hunk 4]
```diff
@@ -4304,33 +4304,82 @@ public function fiscalYearOffset($fieldName) {
...
+   * it is assumed you are using it to get address of a secondary contact.
+   *
+   * @param string $prefix
+   * @param array $extra Additional options.
+   *      Not currently used in core but may be used in override extensions.
    */
-  public function addPhoneFromClause() {
-    // include address field if address column is to be included
-    if ($this->isTableSelected('civicrm_phone')) {
+  public function joinPhoneFromClause($prefix = '', $extra = array()) {
```
Comment (eileenmcnaughton) [inline CRM/Report/Form.php]: Can we change to joinPhoneFromContactClause

File: CRM/Report/Form.php [hunk 5]
```diff
@@ -4304,33 +4304,82 @@ public function fiscalYearOffset($fieldName) {
...
+   * Add Email into From Table if required.
+   *
+   * Prefix will be added to both tables as
+   * it is assumed you are using it to get address of a secondary contact.
+   *
+   * @param string $prefix
+   * @param array $extra Additional options.
+   *      Not currently used in core but may be used in override extensions.
+   */
+  public function joinEmailFromClause($prefix = '', $extra = array()) {
```
Comment (eileenmcnaughton) [inline CRM/Report/Form.php]: Can we change to joinEmailFromContactClause

**Comment by eileenmcnaughton** [comment]
@yashodha there are a few minor things here + it needs squashing. I'm going to squash & put up in an alternate PR with the tweaks. You can then check my reviewers cut is still good & merge

**Comment by eileenmcnaughton** [comment]
I'm closing this in favour of #11814 which is a reviewer's cut of it - note that since that is very big I have started to split off chunks of a few reports so that someone can spend 15-20 minutes testing & confirm a chunk - currently #11867

**Comment by eileenmcnaughton** [comment]
Closing this in favour of #11814 which is a reviewer's cut of this
